File: src/holmes_copilot/copilot.py
from typing import Set, List, Dict, Any


class Copilot:
    def __init__(self) -> None:
        pass

    def generate_rule(events:List[Dict]) -> str:
        repeated_tag_cache = {}
        lines = []
        tag_seq = []
        for event in events:
            holmes_tag = event["holmes-tag"]
            if holmes_tag not in repeated_tag_cache: repeated_tag_cache[holmes_tag] = holmes_tag
            repeated_tag_cache[holmes_tag] += '#'
            holmes_tag = repeated_tag_cache[holmes_tag]
            tag_seq.append(holmes_tag)
        default_rule_name = "_TO_".join(tag_seq).replace("#", "")
        head = default_rule_name + ": sequence "
        repeated_tag_cache = {}
        # calc
        value2fields:Dict[Any, Set[str]] = {}
        for event in events:
            holmes_tag = event["holmes-tag"]
            if holmes_tag not in repeated_tag_cache: repeated_tag_cache[holmes_tag] = holmes_tag
            repeated_tag_cache[holmes_tag] += '#'
            holmes_tag = repeated_tag_cache[holmes_tag]
            for field_name in event:
                if field_name == "holmes-tag": continue
                field_value = event[field_name]
                if field_value not in value2fields: value2fields[field_value] = set()
                value2fields[field_value].add('@'.join([holmes_tag, field_name]))
        # generate groups
        event_groups:Dict[str, List] = {}
        for i, v in enumerate(value2fields):
            # gen one group
            fields = value2fields[v]
            gid = 'g' + str(i)
            if len(set([f.split('@')[0] for f in fields])) == 1: continue
            tag_meet = set()
            for f in fields:
                holmes_tag, field_name = f.split('@')
                if holmes_tag in tag_meet: continue
                tag_meet.add(holmes_tag)
                if holmes_tag not in event_groups: event_groups[holmes_tag] = []
                event_groups[holmes_tag].append((gid, field_name))
        # for
        for tag in tag_seq:
            if tag not in event_groups:
                lines.append('\t[' + tag.replace("#", "") + ']')
                continue
            groups = event_groups[tag]
            constraint = ', '.join(['(' + f + '):' + gid for gid, f in groups])
            lines.append('\t[' + tag.replace("#", "") + '] by ' + constraint)
        lines.insert(0, head)
        return '\n'.join(lines)

if __name__ == "__main__":
    pass

File: java/com/amber/holmes/Copilot.java
package com.amber.holmes;


import java.util.*;

public class Copilot {
    public Copilot() {
    }

    public static String generateRule(List<Map<String, Object>> events) {
        Map<String, String> repeatedTagCache = new HashMap<>();
        List<String> lines = new ArrayList<>();
        List<String> tagSeq = new ArrayList<>();
        
        for (Map<String, Object> event : events) {
            String holmesTag = (String) event.get("holmes-tag");
            repeatedTagCache.putIfAbsent(holmesTag, holmesTag);
            repeatedTagCache.put(holmesTag, repeatedTagCache.get(holmesTag) + "#");
            holmesTag = repeatedTagCache.get(holmesTag);
            tagSeq.add(holmesTag);
        }
        
        String defaultRuleName = String.join("_TO_", tagSeq).replace("#", "");
        String head = defaultRuleName + ": sequence ";
        repeatedTagCache.clear();
        
        // calc
        Map<Object, Set<String>> value2fields = new HashMap<>();
        for (Map<String, Object> event : events) {
            String holmesTag = (String) event.get("holmes-tag");
            repeatedTagCache.putIfAbsent(holmesTag, holmesTag);
            repeatedTagCache.put(holmesTag, repeatedTagCache.get(holmesTag) + "#");
            holmesTag = repeatedTagCache.get(holmesTag);
            
            for (Map.Entry<String, Object> entry : event.entrySet()) {
                String fieldName = entry.getKey();
                if (fieldName.equals("holmes-tag")) continue;
                Object fieldValue = entry.getValue();
                value2fields.putIfAbsent(fieldValue, new HashSet<>());
                value2fields.get(fieldValue).add(holmesTag + "@" + fieldName);
            }
        }
        
        // generate groups
        Map<String, List<AbstractMap.SimpleEntry<String, String>>> eventGroups = new HashMap<>();
        int i = 0;
        for (Object v : value2fields.keySet()) {
            Set<String> fields = value2fields.get(v);
            String gid = "g" + i;
            if (fields.stream().map(f -> f.split("@")[0]).distinct().count() == 1) continue;
            Set<String> tagMeet = new HashSet<>();
            for (String f : fields) {
                String[] parts = f.split("@");
                String holmesTag = parts[0];
                String fieldName = parts[1];
                if (tagMeet.contains(holmesTag)) continue;
                tagMeet.add(holmesTag);
                eventGroups.putIfAbsent(holmesTag, new ArrayList<>());
                eventGroups.get(holmesTag).add(new AbstractMap.SimpleEntry<>(gid, fieldName));
            }
            i++;
        }
        
        // for
        for (String tag : tagSeq) {
            if (!eventGroups.containsKey(tag)) {
                lines.add("\t[" + tag.replace("#", "") + "]");
                continue;
            }
            List<AbstractMap.SimpleEntry<String, String>> groups = eventGroups.get(tag);
            String constraint = groups.stream()
                .map(entry -> "(" + entry.getValue() + "):" + entry.getKey())
                .collect(java.util.stream.Collectors.joining(", "));
            lines.add("\t[" + tag.replace("#", "") + "] by " + constraint);
        }
        lines.add(0, head);
        return String.join("\n", lines);
    }

    public static void main(String[] args) {
    }
}

